File: libs/java-shared/src/main/java/com/ekapasha/shared/response/ErrorResponse.java
package com.ekapasha.shared.response;

import java.time.Instant;
import java.util.List;

public record ErrorResponse(
    int statusCode,
    String error,
    String message,
    Instant timestamp,
    String path,
    String traceId,
    List<ErrorDetail> details) {
  public ErrorResponse(int statusCode, String message, String path) {
    this(statusCode, getErrorLabel(statusCode), message, Instant.now(), path, null, null);
  }

  public ErrorResponse(int statusCode, String message, String path, String traceId) {
    this(statusCode, getErrorLabel(statusCode), message, Instant.now(), path, traceId, null);
  }

  public ErrorResponse(int statusCode, String message, String path, List<ErrorDetail> details) {
    this(statusCode, getErrorLabel(statusCode), message, Instant.now(), path, null, details);
  }

  public ErrorResponse(int statusCode, String message, String path, String traceId, List<ErrorDetail> details) {
    this(statusCode, getErrorLabel(statusCode), message, Instant.now(), path, traceId, details);
  }

  private static String getErrorLabel(int statusCode) {
    return switch (statusCode) {
      case 400 -> "Bad Request";
      case 401 -> "Unauthorized";
      case 403 -> "Forbidden";
      case 404 -> "Not Found";
      case 409 -> "Conflict";
      case 422 -> "Unprocessable Entity";
      case 500 -> "Internal Server Error";
      default -> "Error";
    };
  }
}

File: libs/ts-shared/src/response/ErrorResponse.ts
import { z } from 'zod';

export const ErrorDetailSchema = z.object({
  field: z.string(),
  message: z.string(),
});

export const ErrorResponseSchema = z.object({
  statusCode: z.number(),
  error: z.string(),
  message: z.string(),
  timestamp: z.iso.datetime(),
  path: z.string(),
  traceId: z.string().optional(),
  details: z.array(ErrorDetailSchema).optional(),
});

export type ErrorResponse = z.infer<typeof ErrorResponseSchema>;
export type ErrorDetail = z.infer<typeof ErrorDetailSchema>;

export function createErrorResponse(
  statusCode: number,
  message: string,
  path: string,
  traceId?: string,
  details?: ErrorDetail[],
): ErrorResponse {
  return {
    statusCode,
    error: getErrorLabel(statusCode),
    message,
    timestamp: new Date().toISOString(),
    path,
    traceId,
    details,
  };
}

function getErrorLabel(statusCode: number): string {
  const labels: Record<number, string> = {
    400: 'Bad Request',
    401: 'Unauthorized',
    403: 'Forbidden',
    404: 'Not Found',
    409: 'Conflict',
    422: 'Unprocessable Entity',
    500: 'Internal Server Error',
  };
  return labels[statusCode] || 'Error';
}
